client/src/components/utils/SASCodeDocument.ts

```diff
@@ -0,0 +1,227 @@
+      return `title;footnote;ods _all_ close;
+ods graphics on;
+ods html5${htmlStyleOption} options(bitmap_mode='inline' svg_mode='inline');
+${code}
+;*';*";*/;run;quit;ods html5 close;`;
+    } else {
+      return code;
+    }
+  }
+
+  private wrapSQL(code: string) {
+    return `proc sql;
+${code}
+;quit;`;
+  }
+
+  private wrapPython(code: string) {
+    return `proc python;
+submit;
+${code}
+endsubmit;
+run;`;
+  }
+
+  private wrapCode(code: string): string {
+    let wrapped = code;
+
+    if (this.metadata.languageId === "sql") {
+      wrapped = this.wrapSQL(wrapped);
+    }
+
+    if (this.metadata.languageId === "python") {
+      wrapped = this.wrapPython(wrapped);
+    }
+
+    wrapped = this.wrapCodeWithSASProgramFileName(
+      wrapped,
+      this.metadata.fileName,
+    );
+
+    wrapped = this.wrapCodeWithPreambleAndPostamble(
+      wrapped,
+      this.options?.preamble,
+      this.options?.postamble,
+    );
+    wrapped = this.wrapCodeWithOutputHtml(wrapped);
+
+    return wrapped;
+  }
+
+  // getWrappedCode() returns more code than raw code in editor, and addition code may be added at the beginning or end of raw code.
+  // this method return the position at which the raw code begins after wrapped.
+  private getRawCodeBeginLineNumber(): number {
+    const FRONT_LOCATOR = "LOCATOR-TO-MARK-THE-BEGIN-OF-USER-CODE";
+    const codeWithLocator = FRONT_LOCATOR + this.getRawCode();
+    const wrapped = this.wrapCode(codeWithLocator);
+    return wrapped
+      .split("\n")
+      .findIndex((line) => line.includes(FRONT_LOCATOR));
+  }
+
+  private constructOffsetMap(): void {
+    const selections = this.getSelections();
+
+    let lineNumberInCodeToRun = this.getRawCodeBeginLineNumber();
+
+    this.offsetMap = new Map<LineNumber, Offset>();
+    selections.forEach((selection) => {
+      const { start, end } = selection;
+      for (let i = start.line; i <= end.line; i++) {
+        const offset = {
+          lineOffset: i - lineNumberInCodeToRun,
+          columnOffset: i === start.line ? start.character : 0,
+        };
+        this.offsetMap.set(lineNumberInCodeToRun++, offset);
+      }
+    });
+  }
+
+  private getSelections(): ReadonlyArray<CodeSelection> {
+    // if no valid selection, return whole text as only selection
+    if (!this.selectionsAreEmpty()) {
+      return this.options.selections;
+    } else {
+      const start = { line: 0, character: 0 };
+
+      const codeLines = this.metadata.code.split("\n");
+      const lastLine = codeLines.length - 1;
+      const lastCharacter = codeLines[lastLine].length - 1;
+      const end = { line: lastLine, character: lastCharacter };
+
+      return [{ start, end }];
+    }
+  }
+
+  // priority return selected code, otherwise, return whole code.
+  private getRawCode(): string {
+    return this.metadata.selectedCode.trim() === ""
+      ? this.metadata.code
+      : this.metadata.selectedCode;
+  }
+
+  getWrappedCode(): string {
```

Comment:
Oh, one nitpicky thing I forgot. Since we have `private` modifiers on some of these methods, can we label these as `public` and perhaps move them _before_ the private methods (just where we have a clear idea of that the public interface of SASCodeDocument is)

Reply:
Is this another candidate that should be enforced with eslint? I found [this](https://typescript-eslint.io/rules/member-ordering/) but wasnt sure if we wanted to have a rigid rule for this kind of thing.

Reply:
Done